sdk/core/core/src/az_curl.c

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_curl.h>
+
+#include <az_action.h>
+#include <az_http_header.h>
+
+#include <_az_cfg.h>
+
+/**
+ * Adds a header string to the given CURL list.
+ */
+AZ_NODISCARD az_result
+az_http_header_str_to_curl(struct curl_slist ** const pp_list, char const * str) {
+  AZ_CONTRACT_ARG_NOT_NULL(pp_list);
+  AZ_CONTRACT_ARG_NOT_NULL(str);
+
+  struct curl_slist * const p_list = curl_slist_append(*pp_list, str);
+  if (p_list == NULL) {
+    return AZ_ERROR_HTTP_UNKNOWN;
+  }
+  *pp_list = p_list;
+  return AZ_OK;
+}
+
+AZ_ACTION_FUNC(az_http_header_str_to_curl, struct curl_slist *, az_str_action)
+
+/**
+ * Adds a header to the given CURL list.
+ */
+AZ_NODISCARD az_result az_http_header_to_curl(struct curl_slist ** pp_list, az_pair const header) {
+  AZ_CONTRACT_ARG_NOT_NULL(pp_list);
+
+  return az_span_emitter_to_tmp_str(
```

Comment:
I really don't like how this works because there is not actual implementation here for CURL.
What we see here is that this `az_curl.c` is basically depending on whatever is implemented for `az_span_emitter_to_tmp_str` and `az_http_header_emit_span_seq_action` ... then there is this `az_http_header_str_to_curl_action()`  that somehow get links to `az_http_header_str_to_curl()`

I have to trust that `AZ_ACTION_FUNC` is doing that magic somehow.

Reply:
`Emitter` is not a new abstraction layer. It was already there, it was called sometimes `sequence visitor`, `sequence`, `push iterator`. An action was called a callback. A callback doesn't require to have a pointer on context. An action does.

https://github.com/Azure/azure-sdk-for-c/blob/sergey/refactoring/emitter-design.md

I've updated `az_curl.c` to make it clear what it's doing. IMHO, it should be called `az_curl_slist.c` now. It's very thing adapter which writes directly to `curl_slist`, it's memory safe and handles CURL errors.

Reply:
BTW, you are right, there is not actual implementation here for CURL. Because there is nothing to implement :-D Really, it's not a complex sort algorithm, it's an adapter from one layer to another. Transformation from one data representation into another.

Reply:
yeah, I don't like it.  az_curl_slist.c is not a good name. curl's slist is only for headers.
We will need also code to set a body, and code to get response tokens (headers, code, etc). 
This approach would be only maintainable by functional programming developers :P

Reply:
All abstractions are artificial, they are only a tool for our mind to conquer complexity. In that sense, I feel this rebranding is more than just a rename, it is now actually easier to read and understand. At least, comparing to the two previous incarnations :)